pkg/quantile/agent.go

```diff
@@ -50,14 +54,24 @@ func (a *Agent) Reset() {
 }
 
 // Insert v into the sketch.
-func (a *Agent) Insert(v float64) {
-	a.Sketch.Basic.Insert(v)
+func (a *Agent) Insert(v float64, sampleRate float64) {
+	k := agentConfig.key(v)
+	if sampleRate == 1 {
+		a.Sketch.Basic.Insert(v)
+		a.Buf = append(a.Buf, k)
 
-	a.Buf = append(a.Buf, agentConfig.key(v))
-	if len(a.Buf) < agentBufCap {
-		return
+		if len(a.Buf) < agentBufCap {
+			return
+		}
+	} else {
+		n := uint(math.Ceil(1 / sampleRate))
+		a.Sketch.Basic.InsertN(v, n)
```

Comment:
It strikes me as a bit odd that we'll have different behavior here for dists compared to histograms with regards to `sum` and `count`. Histograms keep the sum calculation as a float and scale using the rate, but use the floor (instead of ceil) for the counts:
https://github.com/DataDog/datadog-agent/blob/486746897294ffe9e74681a30f6283a488b0b8f3/pkg/metrics/histogram.go#L114-L115

It's also interesting that histograms explicitly check for cases where the provided rate == 0, we may want to add a check here as well: https://github.com/DataDog/datadog-agent/blob/486746897294ffe9e74681a30f6283a488b0b8f3/pkg/metrics/histogram.go#L109-L111

Reply:
Ah, I didn't actually look at the histogram code. I can make it consistent.